src/services/jsTyping.ts

```diff
@@ -0,0 +1,271 @@
+
+/* @internal */
+namespace ts.JsTyping {
+
+    interface HostType {
+        directoryExists: (path: string) => boolean;
+        fileExists: (fileName: string) => boolean;
+        readFile: (path: string, encoding?: string) => string;
+        readDirectory: (path: string, extension?: string, exclude?: string[], depth?: number) => string[];
+        writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void;
+    };
+
+    let _safeList: Map<string>;
+    const notFoundTypingNames: string[] = [];
+
+    function tryParseJson(jsonPath: string, host: HostType): any {
+        if (host.fileExists(jsonPath)) {
+            try {
+                // Strip out single-line comments
+                const contents = host.readFile(jsonPath).replace(/^\/\/(.*)$/gm, "");
+                return JSON.parse(contents);
+            }
+            catch (e) { }
+        }
+        return undefined;
+    }
+
+    /**
+     * @param host is the object provoding I/O related operations.
+     * @param fileNames are the file names that belong to the same project.
+     * @param globalCachePath is used 1. to get the safe list file path 2. as cachePath if it is not specified.
+     * @param cachePath is the actual path to get the cached d.ts files from.
+     * @param typingOptions are used for customizing the typing inferring process.
+     * @param compilerOptions are used as a source of typing inferring.
+     */
+    export function discoverTypings(
+        host: HostType,
+        fileNames: string[],
+        globalCachePath: Path,
+        cachePath: Path,
+        typingOptions: TypingOptions,
+        compilerOptions?: CompilerOptions)
+        : { cachedTypingPaths: string[], newTypingNames: string[], filesToWatch: string[] } {
+
+        // a typing name to typing file path mapping
+        const inferredTypings: Map<string> = {};
+
+        if (!typingOptions) {
+            return { cachedTypingPaths: [], newTypingNames: [], filesToWatch: [] };
+        }
+
+        cachePath = cachePath ? cachePath : globalCachePath;
+        // Only infer typings for .js and .jsx files
+        fileNames = fileNames
+            .map(ts.normalizePath)
+            .filter(f => fileExtensionIs(f, ".js") || fileExtensionIs(f, "jsx"));
+
+        const safeListFilePath = ts.combinePaths(globalCachePath, "safeList.json");
+        if (!_safeList && host.fileExists(safeListFilePath)) {
+            _safeList = tryParseJson(safeListFilePath, host);
+        }
+
+        const filesToWatch: string[] = [];
+        // Directories to search for package.json, bower.json and other typing information
+        let searchDirs: string[] = [];
+        let exclude: string[] = [];
+
+        mergeTypings(typingOptions.include);
+        exclude = typingOptions.exclude ? typingOptions.exclude : [];
+
+        if (typingOptions.enableAutoDiscovery) {
+            searchDirs = ts.deduplicate(fileNames.map(ts.getDirectoryPath));

```

Comment:
search directory should also include the projectRootPath if it is set. There are cases when there are no files in the root directory but the package.json and bower.json are located there.